v2/writer.go

```diff
@@ -130,3 +131,65 @@ func (w *Writer) writeIndex(writer io.Writer, carV1 []byte) (int64, error) {
 	// FIXME refactor index to expose the number of bytes written.
 	return 0, index.WriteTo(idx, writer)
 }
+
+// WrapV1File takes a source path to a CARv1 file and wraps it as a CARv2 file
+// with an index, writing the result to the destination path.
+// The resulting CARv2 file's inner CARv1 payload is left unmodified,
+// and does not use any padding before the innner CARv1 or index.
+func WrapV1File(srcPath, dstPath string) error {
+	// TODO: verify src is indeed a CARv1 to prevent misuse.
+	// index.Generate should probably be in charge of that.
+
+	// TODO: also expose WrapV1(io.ReadSeeker, io.Writer),
+	// once index.Generate takes a ReadSeeker.
+
+	// We don't use mmap.Open, so we can later use io.Copy.
+	f1, err := os.Open(srcPath)
+	if err != nil {
+		return err
+	}
+	defer f1.Close()
+
+	idx, err := index.Generate(f1)
+	if err != nil {
+		return err
+	}
+
+	// Use Seek to learn the size of the CARv1 before reading it.
+	v1Size, err := f1.Seek(0, io.SeekEnd)
+	if err != nil {
+		return err
+	}
+	if _, err := f1.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+
+	// Only create the destination CARv2 when we've gathered all the
+	// information we need, such as the index and the CARv1 size.
+	f2, err := os.Create(dstPath)
+	if err != nil {
+		return err
+	}
+	defer f2.Close()
```

Comment:
It seems like f2.Close() is called twice (see line 191)

Reply:
Yep, this is on purpose - duplicate Close calls might be no-ops or might error, but we don't care as we discard the error here anyway. We want to check the close error if everything else went fine, but we also don't want to leak an unclosed f2 if something went wrong, such as the io.Copy error return.

So this defer saves us from the latter - closing the file if we don't get to the end of the func. And since we do that when something went wrong... we don't care about this close error.

Reply:
Makes sense - may be worth adding also a comment with this explanation

Reply:
Sure, will do. I'm kinda used to this pattern at this point, personally, but I agree that a comment can't hurt. I'm not sure what other people do for files they write to, unless they ignore close errors or potentially leak open files :)

Reply:
I use a named return value for the error, but I prefer the method you've used, named return values are bug-prone.

Reply:
Named result parameters can be OK for docs, but big +1 to avoiding "naked returns" and setting results in magic ways.